### DISCUSSION TITLE
How to get allocations in .NET? And how big is an empty array?

### DISCUSSION THREAD

**Discussion by giscus** [linkdotnet/Blog.Discussions#30] [General]
# How to get allocations in .NET? And how big is an empty array?

Often times we hear about allocations on the heap. How can we easily measure this? This article will show you a very easy way of doing so.
Plus we will answer the question how big is an empty array? And if you think 0 bytes, then spoiler alert, that is not the case at all.

https://steven-giesel.com/blogPost/db43d6f4-4b93-415f-be03-600ee358cdfd

**Comment by TimothyMakkison** [comment]
> Used bytes: 24 
> Array has some fields to hold some information like: int Length (4 bytes), long LongLength (8 bytes), int Rank (4 bytes) plus a void* elementType, which is 8 bytes on 64 bit.

Is this correct? I thought empty arrays were 24 bytes with the structure: 
| 8 byte Object Header | 8 byte Method Table | 8 byte Length |

```C#
var before = GC.GetAllocatedBytesForCurrentThread();

var value = new object();
var after = GC.GetAllocatedBytesForCurrentThread();

Console.WriteLine($"Used bytes: {after - before}");
```
This allocates 24 bytes aka | 8 byte Object Header | 8 byte Method Table | 8 byte Pointer |. 
Proving that `GetAllocatedBytesForCurrentThread` must count the method table and header.

I'm not 100% sure but I don't think `Rank` is stored anywhere, instead it looks like multidimensional arrays store each dimensions length and then calculates the rank somehow.

```
Used bytes: 64
Span: [0:8]: [0,0,0,0,0,0,0,0]
Span: [8:16]: [152,88,183,232,251,127,0,0]
Span: [16:24]: [6,0,0,0,0,0,0,0]
Span: [24:32]: [2,0,0,0,3,0,0,0]
Span: [32:40]: [0,0,0,0,0,0,0,0]
Span: [40:48]: [25,0,0,0,50,0,0,0]
Span: [48:56]: [75,0,0,0,100,0,0,0]
Span: [56:64]: [125,0,0,0,150,0,0,0]
```
```C#
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;
unsafe
{
    var before = GC.GetAllocatedBytesForCurrentThread();
    var value = new int[2, 3];
    value[0, 0] = 25;
    value[0, 1] = 50;
    value[0, 2] = 75;
    value[1, 0] = 100;
    value[1, 1] = 125;
    value[1, 2] = 150;

    var after = GC.GetAllocatedBytesForCurrentThread();
    var size = after - before;
    Console.WriteLine($"Used bytes: {size}");

    GCHandle h = GCHandle.Alloc(value, GCHandleType.Pinned);
    IntPtr ptr = h.AddrOfPinnedObject();

    // ugly way of getting the raw objects internals
    ref byte b = ref Unsafe.AsRef<byte>(ptr.ToPointer());
    var off = 16 + 24; // Aka 16 byte header + 8 byte length + 8 byte dimensions + 8 byte padding???
    b = ref Unsafe.Subtract(ref b, off);
    var byteArr = MemoryMarshal.CreateReadOnlySpan(ref b, (int)size);

    for (int i = 0; i < size; i+=8)
    {
        var text = string.Join(',', byteArr.Slice(i, 8).ToArray());
        Console.WriteLine($"Span: [{i}:{i+8}]: [{text}]");
    }
}
```

**Reply by linkdotnet** [reply]
Thanks for checking that. And yes, you're absolutely right. The `Rank` property is calculated:
```csharp
public unsafe int Rank
{
    get
    {
        int rank = RuntimeHelpers.GetMultiDimensionalArrayRank(this);
        return (rank != 0) ? rank : 1;
    }
}

internal static unsafe int GetMultiDimensionalArrayRank(Array array)
{
    int rank = GetMethodTable(array)->MultiDimensionalArrayRank;
    GC.KeepAlive(array); // Keep MethodTable alive
    return rank;
}
```

As you said, it is computed and not stored on the property. The `Length` attribute is stored as property... well in one way or another:
```csharp
public int Length => checked((int)Unsafe.As<RawArrayData>(this).Length);

// This could return a length greater than int.MaxValue
internal nuint NativeLength => Unsafe.As<RawArrayData>(this).Length;

public long LongLength => (long)NativeLength;
```

**RawArrayData** consists of:
```csharp
internal sealed class RawArrayData
{
  public uint Length;
  public uint Padding;
  public byte Data;
}
```

So it is more complicated than I thought initially. By the way your example:
```csharp
var before = GC.GetAllocatedBytesForCurrentThread();

var value = new object();
var after = GC.GetAllocatedBytesForCurrentThread();

Console.WriteLine($"Used bytes: {after - before}");
```
Returns for me 0 bytes. (on .net7 osx-arm64)

**Reply by TimothyMakkison** [reply]
Thanks for the source code, I tried looking at [reference source](https://referencesource.microsoft.com/#mscorlib/system/array.cs) but it's a bit outdated :laughing:

>  By the way your example:
> Returns for me 0 bytes. (on .net7 osx-arm64)

That's strange, the exact same code printed `24 bytes` on .net7 windows release mode, with the structure:

`| Empty header | Method pointer | Null pointer |`

```
Used bytes: 24
Span: [0:8]: [0,0,0,0,0,0,0,0]
Span: [8:16]: [176,147,141,125,252,127,0,0]
Span: [16:24]: [0,0,0,0,0,0,0,0]
```

I guess its a windows vs mac thing or maybe how .net works on mac? I don't know much about cross platform differences.

I wonder if using the object would allocate it?
```C#
var before = GC.GetAllocatedBytesForCurrentThread();

var value = new object();
// Or just box a value.
// var value = (object)123;

var hash = value.GetHashCode();
var after = GC.GetAllocatedBytesForCurrentThread();

Console.WriteLine($"Used bytes: {after - before}");
Console.WriteLine(hash);
```


Love your blog btw, been binging it after reading your FrozenSpan post.

**Reply by linkdotnet** [reply]
I try to reproduce your example and still nothing :D also when I try to run it on [sharplab.io](https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYCMBYAUHgNwEMwACAIwFMAzAezEtIF5SBxAYQDpXLgBBADYDaAYyLBKcAEIBPCQGcAYvXYQwDKMAAqACwZE4ACgCUAbjyESpYgIiMWUSgHdStcgCtKI4CfO5iZDpE8jrM1kS2lNy8ABLBOuy0cJS+lmRE1BJkLBzR/EKi4pKyCspgquqUmrr6RmYWuKiYAJyGACQARACq8pIUcpTyIKQA3hlZpAC0FDT0lAC+HfWNLYZBIWZAA) it yields the same result: o bytes.

Which brings us back to the original question:  Why do we have 24 bytes for an empty array? :D
I found also an [interesting article](https://adamsitnik.com/the-new-Memory-Diagnoser/) from the benchmarkdotnet guys, when they introduced the new `MemoryDiagnoser`. Besides `Array.Empty<int>`, which is 0 bytes allocated (for obvious reason that it is a static readonly used throughout the whole framework), `byte[8]` is exactly 32 byte wide. So again 24 byte for something (on x64).

After fiddling around a bit more I finally found the answer: And for my surprise you were right from the vary starting point. [Sharplab.io](https://sharplab.io/#v2:C4LgTgrgdgPgAgJgAwFgBQBJKBnADgUwGNgA6ACXwENcAKKfAdwAIBLKYAbSQF0BKAbnRY8RUhWp1GTAPYAjAFaiavAUA===) has a nice gimmic for that. These calls
```csharp
Inspect.Heap(new int[0]);
Inspect.Heap(new object());
```

Do have the following memory profile:

![image](https://user-images.githubusercontent.com/26365461/204791420-055d319a-4666-47f6-99db-8697c03d8a07.png)

I will update the blog post later. Thanks again @TimothyMakkison for the super nice and informative chat. Also thanks for the very kind words! Much appreciated! 

Edit: Updated the blog post also crediting you. Thanks again! I learnt a lot.

**Reply by TimothyMakkison** [reply]
> I try to reproduce your example and still nothing :D also when I try to run it on [sharplab.io](https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYCMBYAUHgNwEMwACAIwFMAzAezEtIF5SBxAYQDpXLgBBADYDaAYyLBKcAEIBPCQGcAYvXYQwDKMAAqACwZE4ACgCUAbjyESpYgIiMWUSgHdStcgCtKI4CfO5iZDpE8jrM1kS2lNy8ABLBOuy0cJS+lmRE1BJkLBzR/EKi4pKyCspgquqUmrr6RmYWuKiYAJyGACQARACq8pIUcpTyIKQA3hlZpAC0FDT0lAC+HfWNLYZBIWZAA) it yields the same result: o bytes.

I don't think `Gc.GetAllocatedBytes` works with sharplab.io. Here it should be showing 24 bytes not 0. 

![image](https://user-images.githubusercontent.com/49349513/204807256-00fa1845-0938-4680-826f-ecce2deb42df.png)


> [Sharplab.io](https://sharplab.io/#v2:C4LgTgrgdgPgAgJgAwFgBQBJKBnADgUwGNgA6ACXwENcAKKfAdwAIBLKYAbSQF0BKAbnRY8RUhWp1GTAPYAjAFaiavAUA===) has a nice gimmic for that.

Thanks for showing me this, this is insanely useful, much better than my scuffed `GCHandle.Alloc` trick.


> I will update the blog post later. Thanks again @TimothyMakkison for the super nice and informative chat. Also thanks for the very kind words! Much appreciated!
> Edit: Updated the blog post also crediting you.

Happy to help, learning a lot from your blogs, thanks crediting me!
